### PR TITLE
feat(cb2-7652): implement the test cancellation workflow

### PR DESCRIPTION
**Pull request by tahayazici**
## Cancel a test

- Create a new cancel test component.
- Implement the cancellation service method.
- Update the test result tag to include "Cancelled".

[CB2-7652](https://dvsa.atlassian.net/browse/CB2-7652)

### CODE REVIEW

File: src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.scss
Comment (gjulien-bjss) [inline src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.scss:1]: should delete this file

File: src/app/store/test-records/effects/test-records.effects.ts [hunk 1]
```diff
@@ -118,6 +155,41 @@ export class TestResultsEffects {
     )
   );
 
+  /**
+   * Call PUT Test Results API to cancel test result
+   */
+  cancelTestResult$ = createEffect(() =>
```
Comment (gjulien-bjss) [inline src/app/store/test-records/effects/test-records.effects.ts:161]: it seems strange to me that we need a whole new effect for what is effectively the exact same as an update. Would it now work to subscribe to the value changes of the form, update the editingTestResult and then just dispatch the usual test result update?

File: src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<form [formGroup]="form" (submit)="handleSubmit()">
+  <app-vehicle-header [isEditing]="true" [testResult]="testResult$ | async" [testNumber]="testNumber$ | async"></app-vehicle-header>
+
+  <app-text-area formControlName="reason" name="reason" label="Why do you want to cancel this test?"></app-text-area>
+
+  <button
```
[the remark this answers is not in the document]
Comment (tahayazici) [reply src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.html:6]: This is of `type=submit` which the `<app-button>` cannot replicate.

File: src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.scss
Comment (tahayazici) [reply src/app/features/test-records/amend/views/confirm-cancellation/confirm-cancellation.component.scss:1]: Done

File: src/app/store/test-records/effects/test-records.effects.ts (hunk 1 above)
Comment (tahayazici) [reply src/app/store/test-records/effects/test-records.effects.ts:161]: An example of your suggestion?

But in hindsight, that would go against the dummy component approach. Components, ideally, shouldn't need to how editingTestResult works and how it links to our usual update flow. They should just need to dispatch an action and don't worry about anything else. I would rather have 2 actions that are similar than put any more responsibility on the components.

**Review by owen-corrigan-bjss — DISMISSED** [review]

File: src/app/store/test-records/effects/test-records.effects.ts (hunk 1 above)
Comment (tahayazici) [reply src/app/store/test-records/effects/test-records.effects.ts]: Suggestion applied.

File: src/app/features/test-records/components/vehicle-header/vehicle-header.component.ts [hunk 3]
```diff
@@ -46,8 +48,25 @@ export class VehicleHeaderComponent {
...
+  }
+
+  getTagType(): TagType {
+    if (this.testResult?.testStatus === TestResultStatus.CANCELLED) {
+      return TagType.YELLOW;
+    }
+
+    switch (this.resultOfTest) {
+      case resultOfTestEnum.pass:
+        return TagType.GREEN;
```
[the remark this answers is not in the document]
Comment (tahayazici) [reply src/app/features/test-records/components/vehicle-header/vehicle-header.component.ts]: You're right, I've already modified `get resultOfTest()`. Suggestion applied.

**Review by gjulien-bjss — APPROVED** [review]

**Review by leifkemp-bjss — APPROVED** [review]

**Pull request merged**